server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java

```diff
@@ -1814,7 +1824,7 @@ List<FileRef> findChopFiles(KeyExtent extent, Map<FileRef,Pair<Key,Key>> firstAn
   public synchronized boolean needsSplit() {
     if (isClosing() || isClosed())
       return false;
-    return findSplitRow(getDatafileManager().getFiles()) != null;
+    return isSplitPossible();
```

Comment:
`findSplitRow` checks more conditions than `isSplitPossible`. Would a side-effect of this change be that more tablets are queued up to be split even though they might not need it (see Tablet.importMapFiles())? I wonder if the checks that are performed in `findSplitRow()` should be moved to a place where they are performed after the Tablet is modified, say post-majc, and then the results of that just stored in a variable that is checked when `needsSplit` is called.

Reply:
I agree there are additional checks, and this could lead to more rows being queued up for split than are strictly necessary, but the full check is performed by calling `findSplitRow()` when a split is queued up by calling `needsSplit()`

Reply:
The DatafileManager manages the files for the Tablet. `DatafileManager.datafileSizes` can be modified in several places in DatafileManager. I'm wondering if the call to `FileUtil.findMidPoint` that is currently in `DatafileManager.findSplitRow` should be moved to `DataFileManager` in it's own method. The new method, when invoked, could return a previously calculated split row for the Tablet if the `datafileSizes` object has not been modified. You could use the same pattern with the call to `FileUtil.findLastKey` that is in `DatafileManager.findSplitRow`. This would keep the overall logic the same but provide an optimization in the cases where it's doing FileIO.

Reply:
>  I'm wondering if the call to FileUtil.findMidPoint that is currently in DatafileManager.findSplitRow should be moved to DataFileManager in it's own method.

I experimented with making some changes in the spirit of this, but with a slightly different implementation than described.  Those changes are in [this branch](https://github.com/keith-turner/accumulo/tree/FineAndDandy-3217) which has [this commit](https://github.com/keith-turner/accumulo/commit/5b19850a218ea5149eec4102affa22bd9ca5b23a).  Was trying to achieve the following.

 * Only computes split information when tablets files have changed
 * Avoids doing any computations with the files while holding the tablet lock
 * Caching computation does not prevent splits from happening when user changes config. Like if a tablet was not a candidate for split and then a user lowers the split threshold making it a candidate, then the tablet should split.

I quickly made these changes, so I need to take a second look at them.

Reply:
A good bit of the diff in the commit I linked to are white space changes.  In one method a try catch block was removed because an IOException went away in the code.

Reply:
This call will block for as long as it takes to calculate the new split point. Although no longer in a synchronized block, and will prevent that split point from being calculated more than once. The initial calculation is still going to block the call that triggered the operation. Especially for bulk loads, this can cause bulk load failures that would have otherwise been successes due to timeouts. Since the split happens in another thread, is it that negative to do this there asynchronously from the original triggering call?

Reply:
> Especially for bulk loads, this can cause bulk load failures that would have otherwise been successes due to timeouts. 

Yikes, that is really bad. For the bulk import case, immediately before it calls needsSplit it adds a new file to the tablet, so for that will cause it to always redo the split computation.  Definitely want to avoid the split computations interfering with the bulk import RPC.  Pulling the computation out of any sync block helps with that because it means the bulk import thread will not get stuck on another thread doing the computation. However still need to do something to remove the computation from the bulk import thread itself.

Reply:
Not sure about this, but I am wondering the need split check should just be removed from the bulk import code in tablet.  The bulk import thread has added the file to the tablet and that thread needs to report back that it did that ASAP.  Having the thread do a random table maintenance operation before  reporting back is not ideal.